### DISCUSSION TITLE
Get Profile BackGround

### DISCUSSION THREAD

**Discussion by Amethxst** [isFakeAccount/psnawp#91] [Show and tell]
I apologize if the code will be a bit ugly/confusing. This is for getting the (FHD) background image of the profiles.
(I wrote everything in user.py / ran PSNAWP locally).

The URL would be 'https://image.api.playstation.com/profile/images/acct/prod/'

```
def get_profile_details(self) -> dict[str, Any]:
        """Ottiene dettagli avanzati del profilo dell'utente come l'immagine di sfondo.

        :returns: Un dizionario contenente l'URL dell'immagine di sfondo.

        :raises PSNAWPForbidden: Se il profilo dell'utente è privato e non hai il permesso di vedere i dettagli.
        :raises PSNAWPNotFound: Se l'ID dell'account non esiste.
        """
        try:
            url = f"{BASE_PATH['profile_uri']}{API_PATH['profiles'].format(account_id=self.account_id)}"
            response: dict[str, Any] = self.authenticator.get(url=url).json()  # type: ignore

            
            background_image_url = f"https://image.api.playstation.com/profile/images/acct/prod/{self.account_id}/profile.JPEG?"

            return {
                "backgroundImageUrl": background_image_url
            }
        except PSNAWPForbidden as forbidden:
            raise PSNAWPForbidden(f"Non sei autorizzato a vedere i dettagli del profilo dell'utente {self.online_id}") from forbidden
        except PSNAWPNotFound as not_found:
            raise PSNAWPNotFound(f"Account ID {self.account_id} non esiste.") from not_found
        except Exception as e:
            print("Errore inatteso durante la chiamata API:", e)  # Debug: stampa errori inattesi
            raise e
```

Code to test:

```
import sys
import os

# Aggiungi il percorso della cartella 'src' al PYTHONPATH per utilizzare i file locali
sys.path.insert(0, os.path.abspath("C:\\Users\\meazz\\Desktop\\PSN_APP_RE\\.TEST\\.TEST SFONDO\\psnawp-master\\src"))

from psnawp_api import PSNAWP
from psnawp_api.models.user import User  # type: ignore

def main():
    npsso = "NPSSO CODE"

    psnawp = PSNAWP(npsso)

    online_id = input("ID ONLINE: ")

    try:
        user_example = psnawp.user(online_id=online_id)
    except Exception as e:
        print(f"\nErrore durante l'ottenimento dell'utente per l'Online ID '{online_id}': {e}")
        sys.exit(1)

    try:
        profile_details = user_example.get_profile_details()
        print(f"URL dell'Immagine di Sfondo: {profile_details['backgroundImageUrl']}")
    except Exception as e:
        print(f"\nErrore durante l'ottenimento dei dettagli del profilo per get_profile_details: {e}")

if __name__ == "__main__":
    main()

```

output tested with user 'fec':

```
ID ONLINE: fec
URL dell'Immagine di Sfondo: https://image.api.playstation.com/profile/images/acct/prod/4938611608430181140/profile.JPEG?
```

**Comment by eeews** [comment]
Sometimes that API wont Give you the exact background image the user use. it work on some accounts and some no 
for example 
PSN God: The API will give the background image ( Last background image used )  but psn god doesnt have background image now.
PSN SAR : The background from the api is different from the one he is using RN
PSN Hakoom : The API will return "Not Found" But the psn has a background image if searched the profile on PS4 or PS5

**Comment by Amethxst** [comment]
Just force cacheBuster:

```
def get_profile_details(self) -> dict[str, Any]:
        """Ottiene dettagli avanzati del profilo dell'utente come l'immagine di sfondo.

        :returns: Un dizionario contenente l'URL dell'immagine di sfondo.

        :raises PSNAWPForbidden: Se il profilo dell'utente è privato e non hai il permesso di vedere i dettagli.
        :raises PSNAWPNotFound: Se l'ID dell'account non esiste.
        """
        try:
            url = f"{BASE_PATH['profile_uri']}{API_PATH['profiles'].format(account_id=self.account_id)}"
            response: dict[str, Any] = self.authenticator.get(url=url).json()  # type: ignore

            timestamp = int(datetime.now().timestamp())
            background_image_url = f"https://image.api.playstation.com/profile/images/acct/prod/{self.account_id}/profile.JPEG?cacheBuster={timestamp}"

            return {
                "backgroundImageUrl": background_image_url
            }
        except PSNAWPForbidden as forbidden:
            raise PSNAWPForbidden(f"Non sei autorizzato a vedere i dettagli del profilo dell'utente {self.online_id}") from forbidden
        except PSNAWPNotFound as not_found:
            raise PSNAWPNotFound(f"Account ID {self.account_id} non esiste.") from not_found
        except Exception as e:
            print("Errore inatteso durante la chiamata API:", e)  # Debug: stampa errori inattesi
            raise e
```
without cacheBuster: https://image.api.playstation.com/profile/images/acct/prod/4938611608430181140/profile.JPEG?

with cacheBuster: https://image.api.playstation.com/profile/images/acct/prod/4938611608430181140/profile.JPEG?cacheBuster=1725549086

**Comment by eeews** [comment]
with cacheBuster it worked if the user has background already it will give current background if not it will get the latest background user used  but if the background image url start with  https://image.api.playstation.com/profile/images/acct/prod/

some backgrounds url start https://satchel-cdn.api.playstation.com/generic/permanent/RICHPROFILE/

the https://image.api.playstation.com cant get them

Like PSN Hakoom

if you used https://image.api.playstation.com it will says Not found even with cacheBuster because the background url start with https://satchel-cdn.api.playstation.com

**Comment by Amethxst** [comment]
Yes, actually searching for the mentioned user does not turn up any BGs. May I ask where you got the URL you provided from?

Other problem with cacheBuster = if I remove the BG it still shows the previous one

**Comment by eeews** [comment]
I have got the link from a checker on Discord belong to guy called X22

The link was like that https://satchel-cdn.api.playstation.com/generic/permanent/RICHPROFILE/2019/f91ad/3/2/{account_id}/{image headers}.jpg

also this is background image of an account but from the old communities api when it was working 
"backgroundImage": {
                "sourceUrl": "https://satchel-cdn.api.playstation.com/communities/483876a0-6751-11e5-a4cb-061d9b1a0df3/8a652448-58be-442b-b000-143f78602faa.png",

**Comment by Amethxst** [comment]
I know X22 and his method is undoubtedly more effective

The one I found is the “least worst,” if you can how to apply the URL provided we would have the definitive way

**Comment by eeews** [comment]
He is using the PS4 API that's why he has the ability to fetch The background and Profile Color.

I have a way to get the background image to all accounts but i have to test it through the PS APP and unfortunately i don't have android phone so i can test on it 
i have tried the android simulator but the http toolkit didn't connect successfully to android simulator. 

If you have way which allow me to see the api of ps app on android simulator through http toolkit tell me so i can see if the method will work or nah

**Comment by Amethxst** [comment]
From PS APP I can't view the BG, so I think using http toolkit is almost useless anyway

the use of PS4 API turns out to be more interesting already
